integ/src/updater.rs

```diff
@@ -0,0 +1,213 @@
+const NUM_RETRIES: usize = 5;
+
+// label node with `bottlerocket.aws/updater-interface-version=2.0.0`, so brupop can find right nodes
+// to run updater
+pub async fn label_node(node_names: Vec<String>, kube_config_path: &Path) -> UpdaterResult<()> {
+    // When integration test creates and add nodes to EKS cluster, nodes usually need few mins to be ready.
+    // Label process will be failed if nodes aren't ready. Therefore, add retry strategy here to avoid this situation,
+    // and give nodes more time to be ready.
+    Retry::spawn(retry_strategy(), || async {
+        for node in &node_names {
+            let status = Command::new("kubectl")
+                .args([
+                    "label",
+                    "node",
+                    node,
+                    BRUPOP_NODE_LABEL,
+                    "--kubeconfig",
+                    kube_config_path
+                        .to_str()
+                        .context(update_error::ConvertPathToStr)?,
+                    "--overwrite=true",
+                ])
+                .status()
+                .context(update_error::LabelNode)?;
+            if status.success() {
+                continue;
+            } else {
+                return Err(update_error::Error::BrupopRun);
+            }
+        }
+        Ok(())
+    })
+    .await
+}
+
+// installing brupop on EKS cluster
+pub async fn run_brupop(kube_config_path: &Path) -> UpdaterResult<()> {
+    let brn_status = Command::new("kubectl")
+        .args([
+            "apply",
+            "-f",
+            "yamlgen/deploy/bottlerocket-node-crd.yaml",
+            "--kubeconfig",
+            kube_config_path
+                .to_str()
+                .context(update_error::ConvertPathToStr)?,
+        ])
+        .status()
+        .context(update_error::BrupopProcess)?;
+
+    ensure!(brn_status.success(), update_error::BrupopRun);
+
+    let brupop_resource_status = Command::new("kubectl")
+        .args([
+            "apply",
+            "-f",
+            "yamlgen/deploy/brupop-resources.yaml",
+            "--kubeconfig",
+            kube_config_path
+                .to_str()
+                .context(update_error::ConvertPathToStr)?,
+        ])
+        .status()
+        .context(update_error::BrupopProcess)?;
+
+    ensure!(brupop_resource_status.success(), update_error::BrupopRun);
+
+    Ok(())
+}
+
+fn retry_strategy() -> impl Iterator<Item = Duration> {
+    ExponentialBackoff::from_millis(RETRY_BASE_DELAY.as_millis() as u64)
+        .max_delay(RETRY_MAX_DELAY)
+        .map(jitter)
+        .take(NUM_RETRIES)
+}
+
+// destroy all resources which were created when integration test installed brupop
+pub async fn delete_cluster_resources(kube_config_path: &Path) -> UpdaterResult<()> {
+    // delete namespaces brupop-bottlerocket-aws. This can clean all resources under this namespace like daemonsets.apps
+    let namespace_deletion_status = Command::new("kubectl")
```

Comment:
kube-rs may be an option if you prefer using a library over a command line shell-out.

Reply:
Currently I prefer to use command line shell-out. : ) Or if you highly recommend me to use a library?

Reply:
> Currently I prefer to use command line shell-out. : ) Or if you highly recommend me to use a library?

I generally prefer using a library vs. shelling out when it is expedient to do so for two reasons:
- eliminates a requirement of the runtime environment (i.e. kubectl is required to be present)
- uses cargo and the compiler to understand dependencies (i.e. the dependency on kubectl's behavior and output is "hidden" from cargo and the compiler

That being said, it's not a "strong" recommendation. Sometimes it's a lot easier to shell out to a binary.